### ISSUE TITLE
DataFrameGroupBy object cannot be passed to dict ctor

### ISSUE THREAD

**Issue by mslapek** [pandas-dev/pandas#30621]
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

iris = pd.read_csv("https://raw.githubusercontent.com/pandas-dev/pandas/master/pandas/tests/data/iris.csv")

groupby_obj = iris.groupby('Name')

d = dict(groupby_obj)
```

#### Problem description

Instance of `DataFrameGroupBy` is an iterable yielding pairs (*key*, *frame*),
hence it ought to work with buildin `dict`.

However, it raises error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: attribute of type 'str' is not callable
```

#### Expected Output

`d` should be a dict with groups:

```
{'Iris-setosa':     SepalLength  SepalWidth  PetalLength  PetalWidth         Name
0           5.1         3.5          1.4         0.2  Iris-setosa
1           4.9         3.0          1.4         0.2  Iris-setosa
2           4.7         3.2          1.3         0.2  Iris-setosa
3           4.6         3.1          1.5         0.2  Iris-setosa
...
7          4.6         3.2          1.4         0.2  Iris-setosa
48          5.3         3.7          1.5         0.2  Iris-setosa
49          5.0         3.3          1.4         0.2  Iris-setosa, 'Iris-versicolor':     SepalLength  SepalWidth  PetalLength  PetalWidth             Name
50          7.0         3.2          4.7         1.4  Iris-versicolor
51          6.4         3.2          4.5         1.5  Iris-versicolor
...
146          6.3         2.5          5.0         1.9  Iris-virginica
147          6.5         3.0          5.2         2.0  Iris-virginica
148          6.2         3.4          5.4         2.3  Iris-virginica
149          5.9         3.0          5.1         1.8  Iris-virginica}

```

Notice, it is possible to *workaround* this by use of intermediate
iterable:

```python
d = dict(list(groupby_obj))
```

#### Output of ``pd.show_versions()``

<details>

```
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.0.0-37-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.3
numpy            : 1.17.4
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 42.0.2.post20191203
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

```

</details>

**Comment by TomAugspurger** [comment]
Does it currently not work because the `.keys` attribute of GroupBy?

**Comment by TomAugspurger** [comment]
Deleting `.keys` does seem to fix this.

```python
In [2]: del groupby_obj.keys

In [3]: dict(groupby_obj)
Out[3]:
{'Iris-setosa':     SepalLength  SepalWidth  PetalLength  PetalWidth         Name
 0           5.1         3.5          1.4         0.2  Iris-setosa
 1           4.9         3.0          1.4         0.2  Iris-setosa
 2           4.7         3.2          1.3         0.2  Iris-setosa
 3           4.6         3.1          1.5         0.2  Iris-setosa
 4           5.0         3.6          1.4         0.2  Iris-setosa
 5           5.4         3.9          1.7         0.4  Iris-setosa
 6           4.6         3.4          1.4         0.3  Iris-setosa
 7           5.0         3.4          1.5         0.2  Iris-setosa
...
```

We could deprecate `keys` and use a different name, but IMO it's not worth the churn given the workaround.

**Comment by mslapek** [comment]
Decision to support compatibility with [collections.abc](https://docs.python.org/3/library/collections.abc.html) in various Pandas parts is nontrivial.<sup>1</sup>

---

`help(dict)` mentions ctor overload for `mapping`.
Therefore my first idea was to "convince" Python, that `DataFrameGroupBy`
is not a mapping.

However, it turned out that `dict` does not use *ABCs*:

```python
>>> import collections.abc
>>> isinstance(groupby_obj, collections.abc.Mapping)
False
```

It uses some logic in C (CPython) ([SO reference](https://stackoverflow.com/q/40667093)).

This example shows, that there is unfortunately no other solution than removal of `.keys` (as in [previous comment](https://github.com/pandas-dev/pandas/issues/30621#issuecomment-570578482)).

```python
>>> class Foo:
...     pass
... 
>>> a = Foo()
>>> a.keys = 'foo'
>>> dict(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: attribute of type 'str' is not callable
>>> isinstance(a, collections.abc.Mapping)
False
```

---

<sup>1</sup>(maybe it is worth of separate discussion?)

**Comment by TomAugspurger** [comment]
> Decision to support compatibility with collections.abc in various Pandas parts is nontrivial.1

We've had discussions about that (at least Series being a mapping and a Sequence) elsewhere. No progress on them though.

**Comment by yehoshuadimarsky** [comment]
From for what it's worth, in @wesm's book he provides this useful snippet for converting a GroupBy into a dict:
> `pieces = dict(list(df.groupby('key1')))` 

Here's the link from the GitHub repo https://github.com/wesm/pydata-book/blob/2nd-edition/ch10.ipynb?short_path=6bf65b4#L206

**Comment by mslapek** [comment]
**TL;DR** I postulate to close this issue - to not pollute this repo.

When I've opened this issue, my guess was that it could be resolved without breaking changes. However, it is not the case.

@TomAugspurger Thanks for quick reply after issue opening (< 24h), it shows high quality of `pandas` project.

**Comment by mroeschke** [comment]
Sounds like there isn't further movement or tasks related to this issue. Closing.
